lib/logger/index.js

```diff
@@ -0,0 +1,167 @@
+ * developer will know at a glance what your code expects, and so that they will get a useful
+ * error message if they violate those expectations.
+ *
+ * @type {Object}
+ */
+
+var global = require('../global');
+var ie = require('../ie')
+var cookie = require('../cookie');
+
+// Set the logLevel to whatever the _bvui_debug cookie
+// has stored, otherwise default to INFO
+var logLevel = 1;
+var debugCookie = cookie.read('_bvui_debug');
+var cookieObj = {};
+
+if (debugCookie) {
+  var val;
+  var values = debugCookie.split(',');
+
+  for (var i = 0; i < values.length; i++) {
+    val = values[i].split(':');
+    cookieObj[ val[0] ] = val[1];
+  }
+}
+
+if (cookieObj.logLevel) {
+  logLevel = cookieObj.logLevel;
+}
+
+var Logger = {
+  // Assign these constants so we don't have to remember them.
+  DEBUG: -1, // very noisy
+  LOG: 0,
+  INFO: 1,   // recommended
+  WARN: 2,
+  ERROR: 3,
+  OFF: 4,    // don't log anything
+
+  getLogLevel: function () {
+    return logLevel;
+  },
+  /**
+   * In order to turn off logging, call setLogLevel(Logger.OFF)
+   */
+  setLogLevel: function (level) {
+    logLevel = level || 0;
+    logger('log', [ 'Log level set to', logLevel ]);
+  },
+
+  debug: function () {
+    logLevel < 0 && logger('log', arguments);
+  },
+
+  log: function () {
+    logLevel < 1 && logger('log', arguments);
+  },
+
+  info: function () {
+    logLevel < 2 && logger('info', arguments);
+  },
+
+  warn: function () {
+    logLevel < 3 && logger('warn', arguments);
+  },
+
+  error: function () {
+    logLevel < 4 && logger('error', arguments);
+  },
+
+  count: function () {
+    logLevel < 1 && logger('count', arguments);
+  },
+
+  time: function () {
+    logLevel < 2 && logger('time', arguments);
+  },
+
+  timeEnd: function () {
+    logLevel < 2 && logger('timeEnd', arguments);
+  },
+
+  group: function (maxLevel) {
+    var args = arguments.slice(1);
+    if (logLevel <= maxLevel) {
+      logger('group', args);
+      logger('time',  args);
+    }
+  },
+
+  groupEnd: function (maxLevel) {
+    var args = arguments.slice(1);
+    if (logLevel <= maxLevel) {
+      logger('groupEnd', args);
+      logger('timeEnd',  args);
+    }
+  },
+
+  /**
+   * Assert that an expression is true, and throw the provided message if it is not.
+   *
+   * @param  {Expression} assertion
+   * @param  {String} message
+   */
+  assert: function (assertion, message) {
+    if (logLevel === Logger.OFF) {
+      return;
+    }
+
+    // If we want to assert something computationally expensive, we can pass a function
+    // as the sole argument to Logger.assert. The function should return an array with
+    // two items: the assertion, and the message.
+    if (typeof assertion === 'function') {
+      Logger.assert.apply(null, assertion());
+      return;
+    }
+
+    if (!assertion) {
+      throw new Error('Assertion failed: ' + message);
+    }
+  }
+};
+
+function logger (level, args) {
+  if (logLevel === Logger.OFF) {
```

Comment:
Here we use `Logger.OFF` but above we use explicit values. I think we should use `Logger` properties throughout.